[#197] Klanten API migration command

SonnyBA opened the pull request:
Fixes [#197](https://github.com/maykinmedia/open-klant/issues/197)

**Changes**

Adds a management command which allows the migration of the previous Klanten API to the Klantinteracties API. Two running instances (Klanten and Klantinteracties API) are required to do so with an API token for the Klanten API.

stevenbal commented on src/openklant/management/commands/migrate_to_v2.py:
```diff
@@ -0,0 +1,475 @@
+            try:
+                url_validator(url)
+            except ValidationError as e:
+                return (
+                    f"Invalid URL(s) detected: {str(e.message)}. See "
+                    "migrate_to_v2 --help for correct usage."
+                )
```
We could maybe raise a `CommandError` here to make the command return an error code as well, but it's a minor thing

stevenbal commented on src/openklant/management/commands/migrate_to_v2.py:
```diff
@@ -0,0 +1,475 @@
+
+
+def _retrieve_klanten(url: str, access_token: str) -> Tuple[list[Klant], str | None]:
+    klanten_path = "/klanten/api/v1/klanten"
+
+    _url = urlparse(url)
+    _params = parse_qs(_url.query)
+
+    client = LegacyOpenKlantClient(f"{_url.scheme}://{_url.netloc}", access_token)
```
We could use `furl` here to construct the URL and define the client outside of the `_retrieve_klanten` function and pass that as an argument instead, to avoid having to do it multiple times, but I suppose it's not too much of an issue

SonnyBA replied:
Instantiating the client outside of the `_retrieve_klanten` function would require parsing the query parameters whenever multiple pages exist (pagination) and passing that to the function. Either way we'll have to parse the given URL somewhere.

stevenbal commented on src/openklant/management/commands/migrate_to_v2.py:
```diff
@@ -0,0 +1,475 @@
+    _params = parse_qs(_url.query)
+
+    client = LegacyOpenKlantClient(f"{_url.scheme}://{_url.netloc}", access_token)
+    response_data = client.retrieve(_url.path or klanten_path, params=_params)
```
Is it necessary to pass the `params` in separately? Can it not be part of the first argument?

SonnyBA replied:
We could make it part of the first argument but it would require some extra work, while this currently reuses the `_params` and uses the `params` keyword available from `requests` (and does the work for us).

stevenbal commented on src/openklant/management/commands/migrate_to_v2.py:
```diff
@@ -0,0 +1,475 @@
+        return [], None
+
+    items = response_data.get("results", [])
+    klant_fields = {field.name: field for field in dataclass_fields(Klant)}
```
This could also be defined outside of this function I think (perhaps as a constant)

stevenbal commented on src/openklant/management/commands/migrate_to_v2.py:
```diff
@@ -0,0 +1,475 @@
+
+    logger.info(f"Trying to create {len(klanten)} klanten through the V2 API")
+
+    digitaal_addressen_path = reverse(
+        "klantinteracties:digitaaladres-list",
+        kwargs=dict(version=settings.REST_FRAMEWORK["DEFAULT_VERSION"]),
+    )
+
+    partijen_path = reverse(
+        "klantinteracties:partij-list",
+        kwargs=dict(version=settings.REST_FRAMEWORK["DEFAULT_VERSION"]),
+    )
```
These can also be defined as constants outside of the function I think

stevenbal commented on src/openklant/management/commands/migrate_to_v2.py:
```diff
@@ -0,0 +1,475 @@
+                digitaal_addressen_path, digitaal_adres.dict()
+            )
+
+            if not isinstance(_data, dict):
+                logger.error(
+                    f"Unknown data received creating a digitaal adres: {_data}. "
+                    "Skipping klant."
+                )
+                continue
```
I'm not sure what the user can actually do if some entries fail, but will it be visible in the logs which specific Klanten have failed so that they can be easily identified?

stevenbal commented on src/openklant/management/commands/migrate_to_v2.py:
```diff
@@ -0,0 +1,475 @@
+            )
+            results.extend(_save_klanten(v2_url, klanten))
+
+        return "\n".join(results)
```
I see this returns the results, but since it's a `Command`, is this also part of the output you see when running the command?

stevenbal commented on src/openklant/tests/fixtures/migrate_command/cassettes/digitaal_adres.yaml:
```diff
@@ -0,0 +1,47 @@
+interactions:
```
I think I would put these cassettes under `src/openklants/tests/files/vcr_cassettes/...` instead of `/fixtures`

SonnyBA replied:
I put these in the `migration/cassettes` folder, where there also is an fixture folder (for the V1 setup).

stevenbal commented:
@SonnyBA I haven't worked on this yet, since I've spent most of my time on zgw consumers upgrade for Open Zaak. Could you pick up the feedback for this PR?

joeribekker commented:
Refinement: Waiting for @SonnyBA to get better. If not this week, @annashamray will pick it up.

SonnyBA commented:
Requires https://github.com/maykinmedia/open-klant/pull/280

SonnyBA commented:
@stevenbal I moved code from the management command file to the `migration` folder inside the django project dir, this makes things more readable in my opinion (separate folders for `v1`/`v2` definitions) instead of having everything in one file. I also moved the cassettes to the `migration` folder inside of the project root where there also is a `fixtures` directory (which is used for the docker compose setup).

alextreme commented:
@SonnyBA please round up this PR and release a new version of Open Klant with in

stevenbal commented on docs/development/migration.rst:
```diff
@@ -0,0 +1,49 @@
+Users of Open Klant version ``1.0.0`` willing to migrate to version ``2.0.0`` or higher
+can make use of the ``migrate_to_v2`` management command. This command **only** migrates
+``Klant`` instances from version ``1.0.0``. The command expects two URLs; a URL for the
+Open Klant instance of version ``1.0.0.``and a URL for version ``2.0.0``. The command
```
```suggestion
Open Klant instance of version ``1.0.0`` and a URL for version ``2.0.0`` The command
```

stevenbal commented on docs/development/migration.rst:
```diff
@@ -0,0 +1,49 @@
+.. _migration_index:
+
+Migration
```
I initially read this as documentation for devops, but looking at the lack of docs for Open Klant, I don't think any client/devops will find these docs.

Do we need to document/mention this command in the changelog to make sure at least Sjoerd knows about it?

SonnyBA replied:
I moved the generic part of this file to the user targeted documentation (under `installation`) and also updated the changelog.

stevenbal commented on docs/development/migration.rst:
```diff
@@ -0,0 +1,49 @@
+        ./src/manage.py loaddata migration/credentials migration/test_digitaal_adres
+
+    # Run the corresponding test
+    $ ./src/manage.py test test openklant.tests.test_migrate.MigrateTestCase.test_digitaal_adres
```
```suggestion
    $ ./src/manage.py test openklant.tests.test_migrate.MigrateTestCase.test_digitaal_adres
```

stevenbal commented on src/openklant/management/commands/migrate_to_v2.py:
```diff
@@ -0,0 +1,176 @@
+
+
+def _generate_dummy_token() -> str:
+    token_auth, _ = TokenAuth.objects.get_or_create(application="Migration application")
```
It might be good to remove this token after running the command (and also if the command fails)

SonnyBA replied:
See https://github.com/maykinmedia/open-klant/pull/263/commits/293156ee77fa8ee876d8d0d92053eaf9e6cd1c7f

stevenbal commented on src/openklant/management/commands/migrate_to_v2.py:
```diff
@@ -0,0 +1,176 @@
+
+    logger.info(f"Trying to create {len(klanten)} klanten through the V2 API")
+
+    dummy_token = _generate_dummy_token()
```
If I understand this correctly, it if the list is paginated and `_save_klanten` is run multiple times, you'll have multiple dummy tokens. It might be better to call this outside the loop in `.handle` to avoid this.

SonnyBA replied:
While it used `get_or_create` to retrieve the token before changes were made to this, I agree that querying the database multiple times is unnecessary, see https://github.com/maykinmedia/open-klant/pull/263/commits/293156ee77fa8ee876d8d0d92053eaf9e6cd1c7f.

stevenbal reviewed (changes requested):
Some small questions/remarks, otherwise it looks good :+1:

stevenbal commented on docs/installation/migration.rst:
````diff
@@ -0,0 +1,21 @@
+
+An example of how one might want to run this command can be seen below:
+
+```
````
formatting is off for this:

![image](https://github.com/user-attachments/assets/452afff8-e023-400a-aab0-6efd5120bd36)

you might have to use `.. code-block:: bash`

SonnyBA replied:
Good catch! That fixed it:

![image](https://github.com/user-attachments/assets/4d350361-7f42-46c4-9cc5-ccbdb3519b9b)

stevenbal reviewed (approved):
Only one remark about docs formatting, otherwise good to merge for me :+1:
